### DISCUSSION TITLE
Strategy for Generating Ids when app has own local db offline.

### DISCUSSION THREAD

**Discussion by darkstormdesigns** [ravendb/ravendb#18306] [Q&A]
Hi All

Some advice on this would be great?
I am currently looking at an app that will store data both locally (probably litedb) and in the cloud(ravendb), with some data being generated both on the server and local devices. 
This App will be used in locations (in the middle of forests, etc) where there may not be any data service available, so the need for local storage/creation of data.

This brings up the question of what id strategy is more suitable?

eg:

1. Use the Id with a guid (eg: contact/4f5124b706254abb9dfa28f829f73610).
2. Use the standard Id, but add another field (SyncId and store the guid as a string in that field).

Many Thanks

Ade

**Comment by ayende** [comment]
The easiest way - use a guid or a similar random id. I would recommend using `UUIDv7` for that, because of sortable times.

If you care about readability of the ids, the issue is whatever you have a good way to generate ids for each app instance?
For example, during registration you'll generate a code (such as https://gist.github.com/dgritsko/9554733 ) which should be pretty small unless you have a lot of instances and then use that in the ID.

Mostly, it's about what you care about, RavenDB won't care.

**Reply by darkstormdesigns** [reply]
Ah I see, so I could create my own key pattern as well?
There would be multiple organisations with each having multiple users so something like this....

eg: 0001000166059c7b
0001 - hex for organisation (max 656536)
0001 - hex for user (max 65536)
66059c7b - Seconds Since Epoch in hex

Would be half the size of a guid or UUIDv7.

Many thanks for the pointers...

Ade

**Reply by ayende** [reply]
You _can_ do that, but now you have taken upon yourself the requirement for managing that.
The issue here is that you are now limited to max 1 doc per user per sec, for example.
And overhead of managing users & organizations.

And the end result is an id that doesn't look meaningfully differently from a UUID, in that case, just use UUIDv7 and never thing about those details again

**Reply by darkstormdesigns** [reply]
I don't mind handling the id myself, just looking through different options currently.
I also tried adding the tableprefix eg: 'tablename/0001000166059c7b' and increasing it to milliseconds (increases by 4 characters).
I notice RavenDb uses a sort of tablename prefix (eg: contacts/) for the id. Is there a reason to include something like this or can I just leave it out?

Thanks for your take on this.

Ade
